norddrop/src/device.rs

```diff
@@ -522,22 +521,30 @@ impl NordDropFFI {
     }
 }
 
-fn crate_key_context(
-    logger: slog::Logger,
-    privkey: SecretKey,
-    pubkey_cb: impl Fn(IpAddr) -> Option<PublicKey> + Send + 'static,
-) -> auth::Context {
-    let pubkey_cb = std::sync::Mutex::new(pubkey_cb);
-    let public = move |ip: IpAddr| {
-        let guard = pubkey_cb.lock().expect("Failed to lock pubkey callback");
-        let key = guard(ip)?;
-        drop(guard);
+fn create_key_context(logger: slog::Logger, key_store: Arc<Box<dyn KeyStore>>) -> auth::Context {
+    let privkey = {
+        let key_store = key_store.clone();
+        let logger = logger.clone();
 
-        debug!(logger, "Public key for {ip:?}: {key:?}");
-        Some(key)
+        move || {
+            let privkey: [u8; PUBLIC_KEY_LENGTH] = key_store.privkey().try_into().ok()?;
+            debug!(logger, "Retrieved private key: {:?}", privkey);
+            Some(SecretKey::from(privkey))
+        }
+    };
+
+    let pubkey = {
+        move |ip: IpAddr| {
+            let key_store: Arc<Box<dyn KeyStore>> = Arc::clone(&key_store);
```

Comment:
mega nit: mixing `Arc::clone` and `.clone()`

Reply:
IMHO, this clone is not needed

Reply:
Apparently yes